moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java

```diff
@@ -0,0 +1,193 @@
+            qualifiers.add(annotation);
+          }
+        }
+      }
+      if (qualifiers == null) {
+        qualifiers = Collections.emptySet();
+      }
+      var adapter = moshi.adapter(componentType, qualifiers);
+      var accessor = component.getAccessor();
+      accessor.setAccessible(true);
+      var componentBinding = new ComponentBinding<>(name, jsonName, adapter, accessor);
+      var replaced = bindings.put(jsonName, componentBinding);
+      if (replaced != null) {
+        throw new IllegalArgumentException(
+            "Conflicting components:\n"
+                + "    "
+                + replaced.name
+                + "\n"
+                + "    "
+                + componentBinding.name);
+      }
+    }
+    Constructor<Object> constructor;
+    try {
+      //noinspection unchecked
+      constructor = (Constructor<Object>) rawType.getDeclaredConstructor(constructorParams);
+      constructor.setAccessible(true);
+    } catch (NoSuchMethodException e) {
+      throw new AssertionError(e);
+    }
+    return new RecordJsonAdapter<>(constructor, bindings).nullSafe();
+  };
+
+  private static record ComponentBinding<T>(String name, String jsonName, JsonAdapter<T> adapter, Method accessor) {
+  }
+
+    private final String targetClass;
+    private final Constructor<T> constructor;
+    private final ComponentBinding<Object>[] componentBindingsArray;
+    private final JsonReader.Options options;
+
+    @SuppressWarnings("ToArrayCallWithZeroLengthArrayArgument")
+    public RecordJsonAdapter(
+        Constructor<T> constructor,
+        Map<String, ComponentBinding<?>> componentBindings
+    ) {
+      this.constructor = constructor;
+      this.targetClass = constructor.getDeclaringClass().getSimpleName();
+      //noinspection unchecked
+      this.componentBindingsArray = componentBindings.values().toArray(new ComponentBinding[componentBindings.size()]);
+      this.options = JsonReader.Options.of(componentBindings.keySet().toArray(new String[componentBindings.size()]));
+    }
+
+    @Override
+    public T fromJson(JsonReader reader) throws IOException {
+      var resultsArray = new Object[componentBindingsArray.length];
+
+      reader.beginObject();
+      while (reader.hasNext()) {
+        int index = reader.selectName(options);
+        if (index == -1) {
+          reader.skipName();
+          reader.skipValue();
+          continue;
+        }
+        var result = componentBindingsArray[index].adapter.fromJson(reader);
+        resultsArray[index] = result;
+      }
+      reader.endObject();
+
+      var initialAccess = constructor.canAccess(null);
+      try {
+        if (!initialAccess) {
+          constructor.setAccessible(true);
+        }
+        return constructor.newInstance(resultsArray);
+      } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
+        throw new AssertionError(e);
+      } finally {
+        constructor.setAccessible(false);
```

Comment:
Is there a reason to reset this? I expect the `canAccess` check to not be cheap such that I'm wondering why we don't make it accessible once and never again.

Reply:
Fair enough, I meant to clean this up with the previous cleanup I did with Method

Reply:
Cleaned up with the constructor conversion to MethodHandle